fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/IndexBuilderImpl.java

```diff
@@ -120,148 +117,40 @@ private void rebuild() {
         referenceService.reset();
         membershipService.reset();
 
-        final var txId = UUID.randomUUID().toString();
+        final var reindexService = new ReindexService(persistentStorageSessionManager, objectSessionFactory, ocflIndex,
+                containmentIndex, searchIndex, referenceService, membershipService, membershipPageSize);
+        final ReindexManager reindexManager = new ReindexManager(ocflRepository.listObjectIds(), reindexService,
+                true, reindexBatchSize);
 
+        LOGGER.debug("Reading object ids...");
+        final var startTime = Instant.now();
         try {
-            LOGGER.debug("Reading object ids...");
-
-            try (final var ocflIds = ocflRepository.listObjectIds()) {
-                ocflIds.forEach(ocflId -> {
-                    LOGGER.debug("Reading {}", ocflId);
-                    try (final var session = objectSessionFactory.newSession(ocflId)) {
-                        indexOcflObject(ocflId, txId, session);
-                    } catch (final Exception e) {
-                        // The session's close method signature throws Exception
-                        if (e instanceof RuntimeException) {
-                            throw (RuntimeException) e;
-                        }
-                        throw new RuntimeException(e);
-                    }
-                });
-            }
-
-            containmentIndex.commitTransaction(txId);
-            ocflIndex.commit(txId);
-            referenceService.commitTransaction(txId);
-            indexMembership(txId);
-            LOGGER.info("Index rebuild complete");
-        } catch (final RuntimeException e) {
-            execQuietly("Failed to reset searchIndex", () -> {
-                searchIndex.reset();
-                return null;
-            });
-
-            execQuietly("Failed to rollback containment index transaction " + txId, () -> {
-                containmentIndex.rollbackTransaction(txId);
-                return null;
-            });
-            execQuietly("Failed to rollback OCFL index transaction " + txId, () -> {
-                ocflIndex.rollback(txId);
-                return null;
-            });
-
-            execQuietly("Failed to rollback membership index transaction " + txId, () -> {
-                membershipService.rollbackTransaction(txId);
-                return null;
-            });
-            throw e;
+            reindexManager.start();
+            reindexManager.commit();
```

Comment:
What happens on failure? Isn't this going to leave the db transaction in place?

Reply:
No, [`ReindexManager.commit()`](https://github.com/fcrepo/fcrepo/pull/1783/files#diff-685980790009aa46aa9d033609dbc6da47d661c9f8c251e9b80f6bda4745bdc3R140-R142) calls [`ReindexService.commit()`](https://github.com/fcrepo/fcrepo/pull/1783/files#diff-b834cc79180db3731825fef4c20b6e309d24a06db2379de2799a8c063fdc8204R165) which does the commit and rollback on failure work.

Reply:
Right, but what happens if the call to `reindexManager.start()` throws an exception?

Reply:
Ok I see what you're saying the DB transaction would just sit there.